### PR TITLE
Migrate to use the System.Diagnostics.Metrics API to track internal metrics

### PR DESCRIPTION
**Pull request by suraciii**
Fixes: #7480

This PR migrate most of current `*Statistic`s to `System.Diagnostics.Metrics.Instrument`s, let metrics can be collected directly by common tools (`dotnet-counters`, `OpenTelemetry SDK`, etc.)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/7814)

### CODE REVIEW

**Comment by ReubenBond** [comment]
@suraciii this is fantastic! Is there anything which remains before you mark this PR as "Ready for review"?

**Comment by suraciii** [comment]
> @suraciii this is fantastic! Is there anything which remains before you mark this PR as "Ready for review"?

I'm validating migrated metrics, and considering to improve the way how Orleans itself subscribes its metrics, but it can be done in a further PR.

This PR is ready for review now, you can get all migrated metrics at `src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs`.
Any feedback would be grateful.

File: src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    public const string DIRECTORY_UNREGISTRATIONS_LOCAL = "orleans-directory-unregistrations-local";
+    public const string DIRECTORY_UNREGISTRATIONS_REMOTE_SENT = "orleans-directory-unregistrations-remote-sent";
+    public const string DIRECTORY_UNREGISTRATIONS_REMOTE_RECEIVED = "orleans-directory-unregistrations-remote-received";
+    public const string DIRECTORY_UNREGISTRATIONS_MANY_ISSUED = "orleans-directory-unregistrations-many-issued";
+    public const string DIRECTORY_UNREGISTRATIONS_MANY_REMOTE_SENT = "orleans-directory-unregistrations-many-remote-sent";
+    public const string DIRECTORY_UNREGISTRATIONS_MANY_REMOTE_RECEIVED = "orleans-directory-unregistrations-many-remote-received";
+
+    // ConsistentRing
+    public const string CONSISTENTRING_RINGSIZE = "orleans-consistent-ring-size";
+    public const string CONSISTENTRING_MYRANGE_RINGPERCENTAGE = "orleans-consistent-ring-range-percentage-my";
```
Comment (ReubenBond) [inline src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs:89]:
```suggestion
    public const string CONSISTENTRING_MYRANGE_RINGPERCENTAGE = "orleans-consistent-ring-range-percentage-local";
```

File: src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    // not used
+    public const string DIRECTORY_LOOKUPS_CACHE_ISSUED = "orleans-directory-lookups-cache-issued";
+    // not used
+    public const string DIRECTORY_LOOKUPS_CACHE_SUCCESSES = "orleans-directory-lookups-cache-successes";
+    public const string DIRECTORY_VALIDATIONS_CACHE_SENT = "orleans-directory-validations-cache-sent";
+    public const string DIRECTORY_VALIDATIONS_CACHE_RECEIVED = "orleans-directory-validations-cache-received";
+    public const string DIRECTORY_PARTITION_SIZE = "orleans-directory-partition-size";
+    public const string DIRECTORY_CACHE_SIZE = "orleans-directory-cache-size";
+    public const string DIRECTORY_RING_RINGSIZE = "orleans-directory-ring-size";
+    public const string DIRECTORY_RING_MYPORTION_RINGDISTANCE = "orleans-directory-ring-my-portion-distance";
```
Comment (ReubenBond) [inline src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs:72]:
```suggestion
    public const string DIRECTORY_RING_MYPORTION_RINGDISTANCE = "orleans-directory-ring-local-portion-distance";
```

File: src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    public const string DIRECTORY_RING_MYPORTION_RINGPERCENTAGE = "orleans-directory-ring-my-portion-percentage";
```
Comment (ReubenBond) [inline src/Orleans.Core/Diagnostics/Metrics/InstrumentNames.cs:73]:
```suggestion
    public const string DIRECTORY_RING_MYPORTION_RINGPERCENTAGE = "orleans-directory-ring-local-portion-percentage";
```

File: src/Orleans.Core/Diagnostics/Metrics/Aggregators/CounterAggregator.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+internal class CounterAggregator
+{
+    private static readonly Dictionary<AggregatorKey, CounterAggregator> Aggregators = new();
+
+    private static CounterAggregator FindOrCreate(string name, params KeyValuePair<string, object>[] tags)
+    {
+        var key = new AggregatorKey(name, tags);
+        ref var aggregator = ref CollectionsMarshal.GetValueRefOrAddDefault(Aggregators, key, out var exists);
```
Comment (ReubenBond) [inline src/Orleans.Core/Diagnostics/Metrics/Aggregators/CounterAggregator.cs]: This doesn't look thread-safe - I suggest a lock

Comment (ReubenBond) [reply src/Orleans.Core/Diagnostics/Metrics/Aggregators/CounterAggregator.cs]: My mistake, you removed this already

**Comment by ReubenBond** [comment]
If there are any other changes you'd like to make here, we can do so in a subsequent PR

**Comment by suraciii** [comment]
@ReubenBond Thank you so much for your help with this PR  :)

**Comment by ReubenBond** [comment]
Gladly, @suraciii, your effort here is a tremendous help

**Pull request merged**